src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp

```diff
@@ -0,0 +1,600 @@
+                            });
+
+                if (check_id != node_->check_id)
+                {
+                    node_->check_awaken_count--;
+                    check_id = node_->check_id;
+
+                    if (listener.head())
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            } while (1);
+
+            node_->waiting_count--;
+        }
+
+        /**
+         * Set the caller's 'is_closed' flag (protecting empty_cv_mutex) and
+         * forzes wake-up all listeners on this port.
+         * This function is used when destroying a listener waiting for messages
+         * in the port.
+         */
+        void close_listener(
+                std::atomic<bool>* is_listener_closed)
+        {
+            {
+                std::lock_guard<SharedMemSegment::mutex> lock(node_->empty_cv_mutex);
+                is_listener_closed->exchange(true);
+            }
+
+            node_->empty_cv.notify_all();
+        }
+
+        /**
+         * Removes the head buffer-descriptor from the listener's queue
+         * @param out was_cell_freed is true if the port's cell is freed because all listeners has poped the cell
+         * @throw std::runtime_error if buffer is empty
+         */
+        void pop(
+                Listener& listener,
+                bool& was_cell_freed)
+        {
+            was_cell_freed = listener.pop();
+        }
+
+        /**
+         * Register a new listener
+         * The new listener's read pointer is equal to the ring-buffer write pointer at the registering moment.
+         * @return A shared_ptr to the listener.
+         * The listener will be unregistered when shared_ptr is destroyed.
+         */
+        std::shared_ptr<Listener> create_listener()
+        {
+            return buffer_->register_listener();
+        }
+
+        bool was_check_thread_detached()
+        {
+            return was_check_thread_detached_;
+        }
+
+        /**
+         * Performs a check of the opened port.
+         * When a process crash with a port opened the port can be leave inoperative.
+         * @throw std::exception if the port is inoperative.
+         */
+        void healthy_check(
+                uint32_t healthy_check_timeout_ms)
+        {
+            std::shared_ptr<bool> is_check_ok = std::make_shared<bool>(false);
+
+            was_check_thread_detached_ = false;
+
+            std::shared_ptr<std::mutex> notify_check_done_mutex = std::make_shared<std::mutex>();
+            std::shared_ptr<std::condition_variable> notify_check_done_cv = std::make_shared<std::condition_variable>();
+            std::shared_ptr<bool> is_check_done_received = std::make_shared<bool>(false);
+
+            std::thread check_thread([=]
```

Comment:
Can we change the capture to reference rather than copy?

Reply:
Not in this case. Because is possible that the caller is destroyed before the lambda ends so would not be good to pass references to the caller's local variables.